### ISSUE TITLE
Ability to customise workspace

### ISSUE THREAD

**Issue by IljaDaderko** [decosoftware/deco-ide#30]
It would be nice to have more control over the workspace i.e. I'd love to be able to view/hide components section in order to give my file tree more space, as well as resize console window and perhaps rearrange buttons in main toolbar.

Is this something we can expect in future releases? Great product by the way guys, really looking for this to mature into an awesome ide :)

**Comment by IljaDaderko** [comment]
@dabbott Definitely! If you could kickstart me in the right direction, I can give this a shot over the weekend :)

**Comment by dabbott** [comment]
@IljaDaderko sweet! Ok, here's how it works currently:

The entire layout is defined here in Workspace: https://github.com/decosoftware/deco-ide/blob/master/web/src/scripts/containers/Workspace.jsx

Currently the different regions are hardcoded. Each region is wrapped in a `<Pane />`, which can resize either horizontally or vertically. Resizing dispatches actions, e.g. `setLeftSidebarWidth`, `setLeftSidebarBottomSectionHeight`, which are in turn saved (confusingly) into both our redux store and localstorage (since your layout should persist across app launches). The styles are returned from the method `getStyles`. 

Separate from sizes, there are actions for hiding/showing the left and right sidebar, `setLeftSidebarVisibility`, `setRightSidebarContent`. This happens in WorkspaceToolbar: https://github.com/decosoftware/deco-ide/blob/master/web/src/scripts/containers/WorkspaceToolbar.jsx

*****


Probably the right answer looks more like this:

We have a simple format for the window layout which just specifies positions of boxes using flexbox. Each box has an `id` which specifies what kind of content to show via a hardcoded map `{ id => <Component /> }`. It's stored into localstorage only, and read out upon usage. This can happen from a utils class, e.g. https://github.com/decosoftware/deco-ide/blob/master/web/src/scripts/utils/RecentProjectUtils.js

I made a little pen of what I have in mind: http://codepen.io/dabbott/pen/GqRpxQ?editors=0010

For inspiration, this project does the whole layout thing pretty well: https://github.com/STRML/react-grid-layout. One option would be to just wrap their `<ReactGridLayout />` in a simplified `<WorkspaceLayout />` abstraction layer. My guess is it wouldn't work that well though. Their system is grid based, so it might not be flexible enough (I like being able to resize 1px at a time, instead of thinking about rows and columns.

Other approaches are welcome too. I haven't thought too thoroughly about this.

**Comment by IljaDaderko** [comment]
@dabbott Alright, sounds doable! Will probably need to stretch this over a couple of weekends, but will give it a shot!

**Comment by dabbott** [comment]
@IljaDaderko were you ever able to give this a shot?
